icu4j/main/classes/core/src/com/ibm/icu/impl/personname/PersonNameFormatterImpl.java

```diff
@@ -276,65 +334,47 @@ private PersonNamePattern getBestPattern(PersonNamePattern[] patterns, PersonNam
-                    nameScript = script;
-                }
-            }
-            if (formatterScript == nameScript) {
-                nameLocale = this.locale;
-            } else {
-                // if the name's script is different from the formatter's script, we use addLikelySubtags() to find the
-                // default language for the name's script and use THAT as the name's locale
-                nameLocale = new Locale(ULocale.addLikelySubtags(new ULocale("und_" + UScript.getShortName(nameScript))).getLanguage());
-            }
-            // TODO: This algorithm has a few deficiencies: First, it assumes the script of the string is the script of the first
-            // character in the string that's not COMMON or INHERITED.  This won't work well for some languages, such as Japanese,
-            // that use multiple scripts.  Doing better would require adding a new getScript(String) method on UScript, which
-            // might be something we want.  Second, we only look at the given-name field.  This field should always be populated,
-            // but if it isn't, we're stuck.  Looking at all the fields requires API on PersonName that we don't need anywhere
-            // else.
-        }
-        return nameLocale;
+        return newLocaleWithScript(name.getNameLocale(), getNameScript(name), null);
     }
 
     /**
      * Returns true if the characters in the name match one of the scripts for the specified locale.
      */
-    private boolean nameScriptMatchesLocale(PersonName name, Locale formatterLocale) {
-        // Rather than exhaustively checking all the fields in the name, we just check the given-name
-        // and surname fields, giving preference to the script of the surname if they're different
-        // (we concatenate them into one string for simplicity).  The "name script" is the script
-        // of the first character we find whose script isn't "common".  If that script is one
-        // of the scripts used by the specified locale, we have a match.
-        String nameText = name.getFieldValue(PersonName.NameField.GIVEN, Collections.emptySet())
-                + name.getFieldValue(PersonName.NameField.SURNAME, Collections.emptySet());
+    private boolean nameScriptMatchesLocale(String nameScriptID, Locale formatterLocale) {
         int[] localeScripts = UScript.getCode(formatterLocale);
-        int stringScript = UScript.COMMON;
-        for (int i = 0; stringScript == UScript.COMMON && i < nameText.length(); i++) {
-            char c = nameText.charAt(i);
-            stringScript = UScript.getScript(c);
-        }
+        int nameScript = UScript.getCodeFromName(nameScriptID);
 
         for (int localeScript : localeScripts) {
-            if (localeScript == stringScript) {
+            if (localeScript == nameScript || (localeScript == UScript.SIMPLIFIED_HAN && nameScript == UScript.HAN) || (localeScript == UScript.TRADITIONAL_HAN && nameScript == UScript.HAN)) {
```

Comment:
I don't see it handling Hani or Kana or Kata vs Jpan, etc. Is that handled somewhere else? It looks like the test cases work, but I'm a bit puzzled.

Reply:
`Hana`/`Kata`/`Hani` vs `Jpan` and `Hang`/`Hani` vs `Kore` are dealt with using `UScript.getCode(Locale)`.  If the locale's language is `ja`, this returns an array containing `UScript.KATAKANA`, `UScript.HIRAGANA`, and `UScript.HAN`.  If the locale's language is `ko`, it returns an array containing `UScript.HANGUL` and `UScript.HAN`.  Otherwise (including for Chinese), it falls back on likely subtags.  I had to special-case Chinese because `UScript.getCode(Locale)` doesn't do anything special for it.  (Actually, that's not strictly true-- for `zh-Hant`, it'll return `UScript.HAN` and `UScript.BOPOMOFO`, but that doesn't really help me.)